src/Event/EventFactory.cpp

```diff
@@ -1,4 +1,220 @@
+                if (vehicle->getWorkFactor() >= avgWorkFactor)
+                {
+                    if (vehicle->getCurVehicleStatus() == VehicleStatus::Available
+                    || vehicle->getCurVehicleStatus() == VehicleStatus::Returning)
+                    {
+                        selectedVehicles.push_back(vehicle);
+                        --numVehiclesRequired;
+                    }
+                }
+                if (numVehiclesRequired == 0)
+                {
+                    if (modifyVehicleStatus(selectedVehicles, bitArray))
+                    {
+                        return selectedVehicles;
+                    }
+                    else
+                    {
+                        // Break inner for loop and continue outer while loop
+                        // I know, I know, it's a goto - don't be scared
+                        // This is the easiest way to perform the loop iteration needed and won't harm anyone
+                        // despite what everyone says, these can be useful.
+                        goto tryAgain;
+                    }
+                }
+            }
+            for (const auto & vehicle : orderedVehicles) {
+                if (vehicle->getCurVehicleStatus() == VehicleStatus::Available
+                    || vehicle->getCurVehicleStatus() == VehicleStatus::Returning) {
+                    selectedVehicles.push_back(vehicle);
+                    --numVehiclesRequired;
+                }
+                if (numVehiclesRequired == 0)
+                {
+                    if (modifyVehicleStatus(selectedVehicles, bitArray))
+                    {
+                        return selectedVehicles;
+                    }
+                }
+            }
+            // See, it's friendly :)
+            tryAgain:;
+        }
+    }
+
+    bool EventFactory::modifyVehicleStatus(std::vector<Vehicle *> vehicleList, BitArray *bitArray) {
+        BitArray unmodifiedBitArray;
+        unmodifiedBitArray.setGlobalBitArray(bitArray->getGlobalBitArray());
+        BitArray modifiedBitArray;
+        modifiedBitArray.setGlobalBitArray(bitArray->getGlobalBitArray());
+
+        for (const auto & vehicle : vehicleList)
+        {
+            if (vehicle->getCurVehicleStatus() == VehicleStatus::Available
+            || vehicle->getCurVehicleStatus() == VehicleStatus::Returning)
+            {
+                modifiedBitArray.setGlobalBitArray(modifiedBitArray.modifyBitArray(vehicle->getVehicleID(), true));
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        uint64_t expected = unmodifiedBitArray.getGlobalBitArray();
+        uint64_t desired = modifiedBitArray.getGlobalBitArray();
+        if(bitArray->globalBitArray.compare_exchange_weak(expected, desired))
+        {
+            for (const auto & vehicle : vehicleList)
+            {
+                vehicle->setCurVehicleStatus(VehicleStatus::Responding);
```

Comment:
Maybe document inline as "clean-up"?

Generally, I think this code is quite complex, especially if somebody reading it is not familiar with linearisability and liveness/safety, and could benefit from pointing out gently in a few spots that you know what you are doing. :grin:

Reply:
I agree absolutely. After my final deliverables are submitted I will spend some time cleaning up and commenting up the code.